### PR TITLE
Allow decoration of functions beyond the admin login

### PR DESCRIPTION
**Pull request by MattBlack85**
@kencochrane as discussed in #85

### CODE REVIEW

File: defender/tests.py [hunk 1]
```diff
@@ -724,6 +729,81 @@ def test_disable_username_lockout(self):
         data_out = utils.get_blocked_usernames()
         self.assertEqual(data_out, [])
 
+    @patch('defender.config.BEHIND_REVERSE_PROXY', True)
+    @patch('defender.config.FAILURE_LIMIT', 3)
+    def test_login_blocked_for_non_standard_login_views_without_msg(self):
```
Comment (kencochrane) [inline defender/tests.py]: Can you add docstrings to the test methods explaining what each test does? Not all tests have them now, and I have been meaning to fix that.

File: defender/middleware.py [hunk 2]
```diff
@@ -18,9 +17,10 @@ def __init__(self, *args, **kwargs):
             # `LoginView` class-based view
             try:
                 from django.contrib.auth.views import LoginView
-                watch_login_method = method_decorator(watch_login)
+                our_decorator = watch_login()
+                watch_login_method = method_decorator(our_decorator)
                 LoginView.dispatch = watch_login_method(LoginView.dispatch)
             except ImportError:  # Django < 1.11
-                auth_views.login = watch_login(auth_views.login)
+                auth_views.login = watch_login()(auth_views.login)
```
Comment (kencochrane) [inline defender/middleware.py]: Is that correct? I'm no expert when it comes to decorators so might be correct, it just looks weird with `watch_login()(auth_views.login)`

**Comment by kencochrane** [comment]
It looks like coverage dropped a little, and a few more quantified code minor issues were found. Other than that, pretty good, thanks for the PR.

**Comment by kencochrane** [comment]
I just merged another PR, it looks like there is a conflict.

File: defender/tests.py (hunk 1 above)
Comment (MattBlack85) [inline defender/tests.py]: Sure, no problem

File: defender/middleware.py (hunk 2 above)
Comment (MattBlack85) [inline defender/middleware.py]: That looks weird but it's correct. The old decorator couldn't accept arguments and had a deep of 2. I had to add one more level in order to accept arguments so the first thing the new decorator expects now are the arguments (or the defaults) while here a function was passed in first instance. Does it make sense?

File: defender/tests.py (hunk 1 above)
Comment (kencochrane) [inline defender/tests.py]: Thx

File: defender/middleware.py (hunk 2 above)
Comment (kencochrane) [inline defender/middleware.py]: Yes, thank you.

**Comment by MattBlack85** [comment]
@kencochrane please review again, I added the docstrings but also 2 more commits.
One is excluding test files from coverage which is dropping coverage by 5% but it gives real results now and the latest one is something I discover yesterday while I was testing my PR, we weren't testing against django 1.11, never, becasue we call `python setup.py develop` which was installing `django>=1.8, django<=1.10`; fixed that

**Comment by kencochrane** [comment]
PR looks good, I'll leave it open for a bit in case anyone else wants to review. If no one else comments, I'll merge.

**Comment by MattBlack85** [comment]
:+1:

**Comment by mehtamanan** [comment]
Can I use this yet?

**Pull request merged**